Uri component

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

Follows [our specification](https://github.com/alexandre-daubois/uri-component-specs)

alexandre-daubois committed:
[Uri] Add component

Nyholm commented on src/Symfony/Component/Uri/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.1
```
```suggestion
7.2
```

Nyholm commented on src/Symfony/Component/Uri/QueryString.php:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * @var array<string, string|string[]>
+     */
+    private array $hashMap = [];
```
Let's give this a better name. Like $parameters?

alexandre-daubois replied:
Yes!

Nyholm commented on src/Symfony/Component/Uri/Uri.php:
```diff
@@ -0,0 +1,130 @@
+     */
+    public static function resolve(Uri|string $baseUri, string $relativeUri): string
+    {
+        $baseUri = $baseUri instanceof Uri ? $baseUri : self::parse($baseUri);
```
Micro optimisation: Move this down to line 99.

alexandre-daubois replied:
Good idea

Nyholm reviewed (approved):
Awesome. Looks good.

I've just reviewed the code. I will review the code + spec later.

alexandre-daubois commented:
Thank you Tobias for your review!

I just pushed a few improvements in https://github.com/alexandre-daubois/symfony/pull/11/commits/7bcb8b61972a901f7e9959b569f4ea19553c7537:
- Fixed user and password reencoding when using `Uri::__toString()`
- Added a few more tests
- Implemented fragment text directives [as we specified here](https://github.com/alexandre-daubois/uri-component-specs?tab=readme-ov-file#fragment-text-directives), cc @javiereguiluz as you proposed this feature 🙂 

All the most important features that are "pretty sure" to make it to this hypothetic new component are now implemented. The other ones we mentioned could wait later iterations.

OskarStark commented on src/Symfony/Component/Uri/.gitattributes:
```diff
@@ -0,0 +1,4 @@
+Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitattributes export-ignore
+/.gitignore export-ignore
```
```suggestion
/.git* export-ignore
```
As Nicolas changed that recently

OskarStark commented on src/Symfony/Component/Uri/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+7.2
+---
+
+* Add the component
```
```suggestion
 * Add the component as experimental
```

OskarStark commented on src/Symfony/Component/Uri/README.md:
```diff
@@ -0,0 +1,21 @@
+TypeInfo Component
```
```suggestion
Uri Component
```

alexandre-daubois replied:
Oops 🙃 Thanks

OskarStark commented on src/Symfony/Component/Uri/README.md:
```diff
@@ -0,0 +1,21 @@
+Resources
+---------
+
+* [Contributing](https://symfony.com/doc/current/contributing/index.html)
+* [Report issues](https://github.com/symfony/symfony/issues) and
+  [send Pull Requests](https://github.com/symfony/symfony/pulls)
+  in the [main Symfony repository](https://github.com/symfony/symfony)
```
```suggestion
 * [Contributing](https://symfony.com/doc/current/contributing/index.html)
 * [Report issues](https://github.com/symfony/symfony/issues) and
   [send Pull Requests](https://github.com/symfony/symfony/pulls)
   in the [main Symfony repository](https://github.com/symfony/symfony)
```

OskarStark commented on src/Symfony/Component/Uri/Tests/FragmentTextDirectiveTest.php:
```diff
@@ -0,0 +1,53 @@
+    public static function provideValidFragmentTextDirectives(): iterable
+    {
+        yield [new FragmentTextDirective('start'), ':~:text=start'];
+
```
No need for eh blank lines

OskarStark commented on src/Symfony/Component/Uri/Tests/UriTest.php:
```diff
@@ -0,0 +1,244 @@
+    public static function provideResolveUri(): iterable
+    {
+        yield "Single-level absolute path" => ['https://example.com/foo', '/bar', 'https://example.com/bar'];
+
```
no need for the blank lines

OskarStark commented on src/Symfony/Component/Uri/composer.json:
```diff
@@ -0,0 +1,32 @@
+    "name": "symfony/uri",
+    "type": "library",
+    "description": "Parses URLs",
+    "keywords": [
+        "uri",
+        "url",
+        "symfony"
+    ],
```
AFAIK we keep it on one line

OskarStark reviewed (commented):
The new component should be introduced as experimental

alexandre-daubois commented:
Addressed your comments. Thanks Oskar!

alexandre-daubois commented:
I pushed a critical missing feature, which is the support for **hash maps in query strings** (just like `parse_str()` does). It's been highlighted while leveraging the Uri component in a real world use case. This is now solved in https://github.com/alexandre-daubois/symfony/pull/11/commits/0e22428d90fc0e9972134abe1f2642829d94e0f2.

**IDN hosts support** has been added in https://github.com/alexandre-daubois/symfony/pull/11/commits/205d4c460787b7f0e3f63b79d18b1cae7d230273.

The real world use case can be found here, where I leveraged the component in the Redis cache adapter for demo purpose: https://github.com/alexandre-daubois/symfony/pull/12.

It's starting to look pretty good!

alexandre-daubois commented:
Closing in favor of https://github.com/symfony/symfony/pull/57192. Thank you everyone for your time and review on this draft!
